Implementation of clickable links behaviour for RichTextBox target

mikhail-barg opened the pull request:
Original idea in #13 
Finally a working solution that I'm mostly happy with.
I had to use GUIDs as a placeholders in initially rendered layout to allow using multiple `${rtb-link}` layouts in a same logEvent, or different `${rtb-link}` layouts for different targets.

![working](https://cloud.githubusercontent.com/assets/2208971/11904732/307d2e78-a5d3-11e5-9943-be6871dd9fac.png)


The only caveat is that I have to store additional data in the logEventInfo.Properties dictionary, which makes it potentially visible to client. Should not be a serious problem though.

What worries me is that part of `RichTextBoxLinkLayoutRenderer.Append` which is not thread-safe:
```c#
LinkInfo linkInfo;
object linkInfoObj;
if (logEvent.Properties.TryGetValue(LinkInfo.PropertyName, out linkInfoObj))
{
    linkInfo = (LinkInfo)linkInfoObj;
}
else
{
    linkInfo = new LinkInfo();
    logEvent.Properties.Add(LinkInfo.PropertyName, linkInfo);
}
```
Should I lock on `logEvent.Properties`? Or is there a better way to pass custom data to target's `Write` method from a layout renderer?

One more thing. Using `${where}` layout renderer wrapper to check if a property exists looks really clumsy: `${when:when='${event-properties:item=ShowLink2}'!='':inner=${rtb-link:inner=${event-properties:item=ShowLink2}}}`. Would you approve for creating `${when-property-exists:property=String:inner=Layout}` layout renderer wrapper in the NLog main project?

304NotModified commented:
Cool! I will try to test this weekend. Not sure if I have enough time, otherwise upcoming week. 



> The only caveat is that I have to store additional data in the logEventInfo.Properties dictionary, which makes it potentially visible to client. Should not be a serious problem though.

I don't think this is an issue if it's well documented. Didn't checked the code yet, but it's only one property?


> Should I lock on logEvent.Properties? Or is there a better way to pass custom data to target's Write method from a layout renderer?

I think `lock` would be a good idea. 


> One more thing. Using ${where} layout renderer wrapper to check if a property exists looks really clumsy: ${when:when='${event-properties:item=ShowLink2}'!='':inner=${rtb-link:inner=${event-properties:item=ShowLink2}}}. Would you approve for creating ${when-property-exists:property=String:inner=Layout} layout renderer wrapper in the NLog main project?

Why do we need this? Can't we just write `${rtb-link:inner=${event-properties:item=ShowLink2}}`

304NotModified commented:
note: compare without newline changes: https://github.com/NLog/NLog.Windows.Forms/pull/14/files?w=1

304NotModified commented on NLog.Windows.Forms/FormHelper.cs:
```diff
@@ -1,165 +1,236 @@
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct CHARFORMAT2_STRUCT
```
Please document this really well (and maybe add a link)

304NotModified commented on NLog.Windows.Forms/FormHelper.cs:
```diff
@@ -1,165 +1,236 @@
+        /// <param name="hyperlink">value of the new link</param>
+        internal static void ChangeSelectionToLink(RichTextBox textBox, string text, string hyperlink)
+        {
+            int selectionStart = textBox.SelectionStart;
```
please remove old code and document some about this rtf1 (because you understand this :))

304NotModified commented on NLog.Windows.Forms/FormHelper.cs:
```diff
@@ -1,165 +1,236 @@
+            //textBox.Select(position + text.Length + hyperlink.Length + 1, 0);
+        }
+
+        private static void SetSelectionStyle(RichTextBox textBox, UInt32 mask, UInt32 effect)
```
We need also some docs on this, because it's a bit difficult

304NotModified commented on NLog.Windows.Forms/RichTextBoxLinkLayoutRenderer .cs:
```diff
@@ -0,0 +1,83 @@
+            object linkInfoObj;
+            if (logEvent.Properties.TryGetValue(LinkInfo.PropertyName, out linkInfoObj))
+            {
+                linkInfo = (LinkInfo)linkInfoObj;
```
I would recommend a 'as' instead of cast

mikhail-barg replied:
I'm not sure on this. Explicit cast would at least fail in (unlikely) case somebody have added something else with this key.

304NotModified commented on NLog.Windows.Forms/RichTextBoxLinkLayoutRenderer .cs:
```diff
@@ -0,0 +1,83 @@
+        {
+            internal const string PropertyName = "NLog.Windows.Forms.RichTextBoxLinkLayoutRenderer.LinkInfo";
+
+            private readonly object m_lock = new object();
```
please no Hungarian names

mikhail-barg replied:
Fixed.

304NotModified commented:
Please rebase, for the newline changes. Sorry that this was resulting in a conflict.

mikhail-barg commented:
No problem, I managed to merge without rebase.

mikhail-barg commented:
Added comments, also links to original implementation from CodeProject for displaying links with hidden text.
Added locks around logEvent.Properties.

> Can't we just write `${rtb-link:inner=${event-properties:item=ShowLink2}}`

Thanks for the proposal! I've added explicit check for empty inner layout to `${rtb-link}`

304NotModified commented:
Great! 

Can we add some more unit tests? That will guarantee that this code will keep working after a few years.

mikhail-barg commented:
Sure, I just have to find some courage and add tests for both this and previous changes.

304NotModified commented:
OK! 

If you add some tests unrelated to this PR, please send a separate PR for that. (and of you only add tests in the PR, you can merge it by yourself if you like . ) 

PS : nice that the compare is now clear without newline change noise :)

mikhail-barg commented:
Well, I'm still struggling to make unit-tests run in my VS2015Community. 
What I tried:
* Installing `xUnit.net Test Extensions` (https://github.com/xunit/testextensions.xunit). It installed, but I still don't see any unit-tests in the VS Test Explorer (reboot, rebuilt).
* Installing `xUnit.net runner for Visual Studio` (https://visualstudiogallery.msdn.microsoft.com/463c5987-f82b-46c8-a97e-b1cde42b9099). Does not install because of some NPE.
* Upgrading xUnit reference of the project to v2.x (as you proposed in another thread). Did not upgrade with the following error:
`Could not install package 'xunit.assert 2.0.0'. You are trying to install this package into a project that targets '.NETFramework,Version=v3.5', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.`

Could you please help me with this?

304NotModified commented:
Sure! It's important that unit tests works locally!. 

I would recommend xunit 2, we use it a lot in other nlog projects. First upgrade the test project   to dotnet 4.5 (as xunit2 won't support older dotnet versions), then upgrade the xunit package. You need also the xunit2 visual studio runner I guess. 

Let me know if you get it working. Otherwise we try something else. 

I use resharper and then everything works out of the box.

304NotModified commented:
You can try 

https://www.jetbrains.com/buy/opensource/?product=resharper

Just refer to this repository and the main website.  

I cannot live without resharper :D

304NotModified commented:
If you do try to apply for resharper, maybe I should add you first to the NLog team. Just let me know.

mikhail-barg commented:
> First upgrade the test project to dotnet 4.5 (as xunit2 won't support older dotnet versions), then upgrade the xunit package. 

It slipped off my mind that I can upgrade just the test project. Still by itself it just did not help.

> You need also the xunit2 visual studio runner I guess.

I've created a project to try running xunit, and adding `xunit.runner.visualstudio` to the project itself does work. But it gets to `packages.config` and project references, which you currently don't have.

Still I found that you have it in `NLog.UnitTests/packages.config`, so I think I'll go that way.

mikhail-barg commented:
Regarding ReSharper — it's very generous of you to suggest adding me to NLog core team just to allow me applying for it, thanks! But I'd rather be invited into the team for other reasons, like skills ) 

Also I prefer to not rely on tools that I can't use on daily basis at my main workplace. So for now I'll try to figure things out without ReSharper. Thanks again for your kind proposal!

mikhail-barg commented:
Hm, I've managed to run the tests, but I get 4 of them failing. Which is really weird as I have no differences (except project settings) from the version that is currently in this PR and has `AppVeyor build succeeded`.

The tests are failing with specific RTF checks like 'Assert.True(result.Contains(@"\cf0\highlight0\f1\par"));' If I check the RTF myself I see `\cf0\highlight0\par` no f1 tag. Could it be that I have another RTB ontrol version because of my Win10 OS?

mikhail-barg commented:
Another thing is, would you like to see the upgrade to xUnit 2 in a separate PR, or it's OK to keep it here?

304NotModified commented:
> Still I found that you have it in NLog.UnitTests/packages.config, so I think I'll go that way.

Yes it's in the packages.config in the test project. 

> Regarding ReSharper — it's very generous of you to suggest adding me to NLog core team just to allow me applying for it, thanks! But I'd rather be invited into the team for other reasons, like skills )

Well then you got it wrong ;) As I said before, I was impressed with your contributions and good communication - that was the reason I added you as collaborator. I would never include someone to the team solely for Resharper. But I have to admit that it wasn't clear in that. Because of your nice contributions, I really would like it if you join NLog! 

> Also I prefer to not rely on tools that I can't use on daily basis at my main workplace. So for now I'll try to figure things out without ReSharper. Thanks again for your kind proposal!

Good point, but if your daily job involves programming in C#, you should use at least Resharper or an alternative IMO. It's a time saver and improves code in general.   

> Hm, I've managed to run the tests, but I get 4 of them failing. Which is really weird as I have no differences (except project settings) from the version that is currently in this PR and has AppVeyor build succeeded.
> 
> The tests are failing with specific RTF checks like Assert.True(result.Contains(@"\cf0\highlight0\f1\par")); If I check the RTF myself I see \cf0\highlight0\par, no f1 tag. Could it be that I have another RTB ontrol version because of my Win10 OS?

I works on AppVeyor, but fail locally? Which ones?



> Another thing is, would you like to see the upgrade to xUnit 2 in a separate PR, or it's OK to keep it here?

A new PR is preferred, but if it's already in this PR, it's not that big deal in this case, because we are almost ready to merge this one.

mikhail-barg commented:
And one more question. I'd like to check if the dictionary of the target gets links removed when they are deleted from the control. But the dictionary is private. Do you have any policy to access class private members? Like `Microsoft.VisualStudio.TestTools.UnitTesting.PrivateObject`?

304NotModified commented:
> And one more question. I'd like to check if the dictionary of the target gets links removed when they are deleted from the control. But the dictionary is private. Do you have any policy to access class private members? Like Microsoft.VisualStudio.TestTools.UnitTesting.PrivateObject?

You like to check it in your unit tests?
We would make it `internal` if it's a property or method. (fields get wrapped), and use the `InternalsVisibleToAttribute `

mikhail-barg commented:
> Because of your nice contributions, I really would like it if you join NLog!

I'm very pleased by your warm words, thanks!

> works on AppVeyor, but fail locally? Which ones?

CustomRowColoringTest, CustomWordRowColoringTest, NoColoringTest and SimpleRichTextBoxTargetTest because of
`Assert.True(result.Contains(@"\cf0\highlight0\f1\par"));`

> A new PR is preferred. But if it's already in this PR, it's not that big deal in this case because we are almost ready to merge this one.

OK, I'll just add the tests that are related to the link functional. Others would go to another PR then.

304NotModified commented:
> CustomRowColoringTest, CustomWordRowColoringTest, NoColoringTest and SimpleRichTextBoxTargetTest because of
> Assert.True(result.Contains(@"\cf0\highlight0\f1\par"));

I will check at my machine also. 



> OK, I'll just add the tests that are related to the link functional. Others would go to another PR then.

sounds good :)

304NotModified commented:
> CustomRowColoringTest, CustomWordRowColoringTest, NoColoringTest and SimpleRichTextBoxTargetTest because of
> Assert.True(result.Contains(@"\cf0\highlight0\f1\par"));
> I will check at my machine also.

Those also fail with me with .net 45 (xunit1)

mikhail-barg commented:
Ok, I've added few tests that should cover most of the troublesome parts. I've added `assembly:InternalsVisibleTo` attribute and exposed `RichTextBoxTarget.LinkedEventsCount` property. To make InternalsVisibleTo attribute work I had to sign test assembly, hope it would not cause any problems.

What I was not able to test is actual link clicking behaviour. I've tried using `mouse_event` calls as proposed [here](http://stackoverflow.com/a/8273118/376066), but got only cursor over the link and no actual clicking. Not sure if it could be made to work at all.

BTW, I don't see codecov digest for the last commits. Should I fire it somehow?

304NotModified commented:
> Ok, I've added few tests that should cover most of the troublesome parts. I've added assembly:InternalsVisibleTo attribute and exposed RichTextBoxTarget.LinkedEventsCount property. To make InternalsVisibleTo attribute work I had to sign test assembly, hope it would not cause any problems.

That's no problem, as we won't release the test assembly :)



> What I was not able to test is actual link clicking behaviour. I've tried using mouse_event calls as proposed here, but got only cursor over the link and no actual clicking. Not sure if it could be made to work at all.

That would be cool if it worked.

> BTW, I don't see codecov digest for the last commits. Should I fire it somehow?

It's updating the codecov post. It was first 40%, now almost 70% :D

mikhail-barg commented:
Well, with a bit of black magic I've added an actual link click test. I think that should be it.

304NotModified commented on NLog.Windows.Forms.Tests/RichTextBoxTargetTests.cs:
```diff
@@ -588,13 +582,234 @@ public void ManualRegisterTestWithRetention()
+            string resultRtf = ExtractRtf(target.TargetRichTextBox);
+            string resultText = target.TargetRichTextBox.Text;
+            Assert.False(Regex.IsMatch(resultRtf, @"(\([a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}\))"));  //the placeholder GUID was replaced
+            //cant check for specific ids here because of the possible parallel execution
```
Maybe we should disable parallel execution?

mikhail-barg replied:
Actually, I'm not sure that this is because of parallel execution. It could be just that executor is not reloading classes, so static fields are not re-initialized. Anyway I was able to test all that I thought was necessary.

304NotModified replied:
OK, then I think it's better to remove those comments, and the look as a TODO. (and we can disable parallel execution). https://xunit.github.io/docs/running-tests-in-parallel.html 

I used `CollectionAttribute` before, as I don't think parallelism in unit tests is important -  I like more a robust outcome.

mikhail-barg replied:
Removed comments.

Regarding the side-effects of the parallel or sequential test execution - I'm still not sure what is causing the ID shift, but anyway it seems that relying on specific `SequenceID`s is not a great idea. So current checks should be OK too.

304NotModified commented on NLog.Windows.Forms.Tests/packages.config:
```diff
@@ -1,5 +1,11 @@
 <packages>
-  <package id="NLog" version="4.0.0" targetFramework="net35" />
-  <package id="xunit" version="1.9.2" targetFramework="net35" />
+  <package id="NLog" version="4.2.2" targetFramework="net45" />
```
Do we need this version of NLog? Isn't it 4.0 compatible?

mikhail-barg replied:
Sorry, it got there when I was fighting with xUnit versions.
Reverting to Nlog 4.0.1. Version 4.0.0 is not able to find the layout renderer.

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -299,6 +318,100 @@ public RichTextBoxTargetMessageRetentionStrategy MessageRetention
+        /// <summary>
+        /// Used to capture link placeholders in <see cref="SendTheMessageToRichTextBox"/>
+        /// </summary>
+        private readonly Regex linkAddRegex = new Regex(@"(\([a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}\))", RegexOptions.Compiled);
```
I would recommend to use regex caching instead of compiled ones.

mikhail-barg replied:
I was not aware about caching functional, so that you for pointing this out. I've read (Compilation and Reuse in Regular Expressions)[https://msdn.microsoft.com/en-us/library/8zbs0h2f(v=vs.110).aspx] and (Best Practices for Regular Expressions in the .NET Framework)[https://msdn.microsoft.com/en-us/library/gg578045(v=vs.110).aspx] and I still don't think there's a reason to not use compiled regexes in this case.

if I get the text right, the regex cache is just 15 items. So it looks like a valuable user-space resource. And consuming 20% of it for some helper library code looks like a waste. I think well-behaving libraries should not do such things.

I think that these regexes are quite small, so the memory overhead should be rather small too, as well as the compilation time.

Still, please make a final decision.

304NotModified replied:
I have seen a lot of memory issues with compiled regexes. I have read a lot of it in the past, and I did re-read your links. Also find this useful: http://stackoverflow.com/questions/9969158/when-not-to-use-regexoptions-compiled

I think the we can use compile, iif:

- The regex is compiled, eg. use `static` (and preferred, lazy)
- The regex is used a lot. This is the case for 'linkAddRegex', and 'linkRemoveRtfRegex', but not for `linkClickRegex`
- If performance is important, don't use regexes if that's is possible
- and of course prevent backtracking. 

So we can keep two of them, if there will be static.

mikhail-barg replied:
Made these two regexes static, added lazy initialization (also was thinking of it, but forgot).

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -454,6 +577,47 @@ private void AttachToControl(Form form, RichTextBox textboxControl)
+            }
+
+            int id;
+            if (!Int32.TryParse(match.Groups[2].Value, out id))
```
Is there a reason for int32 instead of int?

mikhail-barg replied:
Updated.

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -299,6 +318,100 @@ public RichTextBoxTargetMessageRetentionStrategy MessageRetention
+        /// <summary>
+        /// Used to parse links in <see cref="TargetRichTextBox_LinkClicked"/>
+        /// </summary>
+        private readonly Regex linkClickRegex = new Regex(@"(.*)#" + LINK_PREFIX + @"(\d+)", RegexOptions.Compiled);
```
the `.*` makes this regex really a performance eater? (a lot of backtracking)

mikhail-barg replied:
Moved regex to actual usage place, removed `.*` part.

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -299,6 +318,100 @@ public RichTextBoxTargetMessageRetentionStrategy MessageRetention
+        /// <summary>
+        /// Internal prefix that is added to the link id in RTF
+        /// </summary>
+        private const string LINK_PREFIX = "link";
```
I think caps and snake case is against the C# naming guidlines. I prefer LinkPrefix

mikhail-barg replied:
Updated

304NotModified commented:
I think also. I will do some last checking.

I think it would be nice to write a news post (nlog-project.org) about this cool feature, what do you think?

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -454,6 +584,48 @@ private void AttachToControl(Form form, RichTextBox textboxControl)
+            if (linkClickEvent != null)
+            {
+                string linkText = e.LinkText.Substring(0, match.Index);
+                linkClickEvent(this, match.Groups[1].Value, logEvent);
```
linkText is unused. Is this a bug?

mikhail-barg replied:
Yes! Thanks for finding it! I wonder, why I haven't noticed this myself, while we have "Treat warnings as errors" set to "All". Anyway, I've fixed this.

mikhail-barg replied:
Updated unit test to check for this explicitly.

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -299,6 +318,107 @@ public RichTextBoxTargetMessageRetentionStrategy MessageRetention
+                {
+                    lock (linkRegexLock)
+                    {
+                        if (linkAddRegex == null)
```
This is a bit tricky IMO. I think this lazy init should be encapsulated in a property. Now we can use `linkAddRegex` and `linkRemoveRtfRegex` too soon (theoretical), which lead to unexpected behavior / nullref ex.

When encapsulating in a property this is also possible, but we can find easily all usages of the field (and can be replaced with the property).

Do you agree?

mikhail-barg replied:
Well, I don't like turning `linkAddRegex` and `linkRemoveRtfRegex` into properties with lazy initialization. It wold look like that, right:

```c#
private static Regex LinkAddRegex 
{
   get 
   {
       lock (linkRegexLock) 
       {
           if (linkAddRegex == null)
           {
              linkAddRegex = new Regex(...);
           }
       }
       return linkAddRegex;
   }
}
```

So we would have to not only check if the regex actually exists, but also synchronize this check. And it would be inter-instance synchronization because of `static` fields nature. And this would happen for every log event.

Current code guards access to these two regexes with `if (SupportLinks)` which assures creation of regexes. (I'll update the code to assure the underlying `supportLinks` field is set after the regex initialization.) So all the locking happens on rarely called property change.

I also don't like (costy) regex initialization to happen during actual log event, so I would like to keep initialization in the `SupportLinks` setter anyway, which lessens the value of having lazily-initialized properties.

Maybe we could just add better comments to this regexes usage? Or remove lazy initialization as whole?

304NotModified commented:
I tested it and check the code. Looks nice and works well!  I have send a PR for your PR for naming / code style - just try to keep everything consistent (and C# naming rules in my mind)

Also found 2 potential code issues. (unused var, lazy order). If you fix those, then I will merge and release it :)

mikhail-barg commented:
Thanks for PR and comments! Especially for finding unfortunate `linkText` problem. I've updated everything except for the lazy regex initialization. Please see my comment on this and make a final decision.

Looking forward for a release :), can't wait to use in our code here!

Regarding a news post — yes, I agree. Being a NLog user I'd like to know about such things, and I think it worth notifying everybody (hope I does not sound too self-proud).

304NotModified commented:
> Thanks for PR and comments! Especially for finding unfortunate linkText problem.

Thanks to Resharper ;)

Maybe we are missing a unit test? Because the tests were succeeding?

> I've updated everything except for the lazy regex initialization. Please see my comment on this and make a final decision.

You have an excellent point about the performance. So better comments is the way, and I see you did that. 

> Looking forward for a release :), can't wait to use in our code here!

Will do that this week
 
> Regarding a news post — yes, I agree. Being a NLog user I'd like to know about such things, and I think it worth notifying everybody (hope I does not sound too self-proud).

Well I think this is a really cool feature, so good reason to be proud of. 

Do you like to write the news post? I think that's best from the creator  :)
See https://github.com/NLog/NLog.github.io#add-a-news-post-

The new release will be 4.2 (semver)

mikhail-barg commented:
> Maybe we are missing a unit test? Because the tests were succeeding?

Yes. I've updated the test to check this too in a previous commit.

> Do you like to write the news post? I think that's best from the creator :)
> See https://github.com/NLog/NLog.github.io#add-a-news-post-

Thanks, I'll try :) Am I right that I need to create a PR for this, and you'll merge it in when the release is available on NuGet?

I'll also update target's wiki on new properties.

304NotModified commented:
Yes your right about the PR and merge :)

mikhail-barg commented:
One more thing regarding blog post. I'd like to add some images of textbox with ling to the post. Is it OK if I put them into NLog.github.io/images/posts/? Is it OK to add images at all?

304NotModified commented:
Images are a good idea! You can also (ab) use the upload functionality in the github issues. 

Otherwise a sub folder (inside images)  for the post is preferred.

mikhail-barg commented:
Thanks! And another question. What date to use for the post? Or could you change it yourself when merging PR after release?

304NotModified commented:
I will sync the date when releasing. :)

mikhail-barg commented:
I've created a PR with a post (https://github.com/NLog/NLog.github.io/pull/47). Also updated target's wiki, and added a page for the layout renderer.

mikhail-barg commented:
I've also created a small sample project to make screenshots. Would you like to add it to the solution?

mikhail-barg commented:
Attaching test app in case it's needed later.
[NLog.Windows.Forms.TestLinksApplication.zip](https://github.com/NLog/NLog.Windows.Forms/files/200676/NLog.Windows.Forms.TestLinksApplication.zip)
